File: src/fields/flattened_type_relation.cpp
#include <algorithm>
#include <iterator>
#include "fields/flattened_type_relation.h"
#include "fields/null_terminated_array.h"

FlattenedTypeRelation::FlattenedTypeRelation(FlattenedType* ft, const std::vector<FieldLinkDefinition*>& fls) {
    flattenedType = ft;

    std::map<int, std::vector<FieldLinkDefinition*>*> groupedByOriginFD;

    // Group FieldLinkDefinitions by their origin field definitions' IDs
    for (const auto& fl : fls) {
        std::vector<FieldLinkDefinition*>* list = groupedByOriginFD[fl->getOriginFD()->getId()];
        if(list == nullptr) {
            list = new std::vector<FieldLinkDefinition*>();
            groupedByOriginFD[fl->getOriginFD()->getId()] = list;
        }
        list->push_back(fl);
    }

    // Initialize fieldLinks to the correct size
    fieldLinks = new FieldLinkDefinition**[flattenedType->getNumberOfFields()];

    // Fill fieldLinks with the grouped FieldLinkDefinitions
    for (int i = 0; i < flattenedType->getNumberOfFields(); ++i) {
        FieldDefinition** fdArray = flattenedType->getFieldsReverse()[i];

        int j = 0;
        while (fdArray[j] != nullptr) {
            FieldDefinition* fd = fdArray[j];
            j++;

            std::vector<FieldLinkDefinition*>* fls = groupedByOriginFD[fd->getId()];

            if(fls != nullptr) {
                fieldLinks[i] = nullTerminatedArrayFromVector<FieldLinkDefinition*>(*fls);
            }
        }
    }
}

void FlattenedTypeRelation::followLinks(Direction* direction, Object* relatedObj, Field* field) {
    FieldLinkDefinition** flArray = fieldLinks[field->getId()];

    int i = 0;
    while (flArray[i] != nullptr) {
        FieldLinkDefinition* fl = flArray[i];
        i++;

        direction->transmit(field, fl, relatedObj);
    }
}

File: java/fields/src/main/java/network/aika/type/FlattenedTypeRelation.java
package network.aika.type;

import network.aika.fields.defs.FieldDefinition;
import network.aika.fields.direction.Direction;
import network.aika.fields.field.Field;
import network.aika.fields.defs.FieldLinkDefinition;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class FlattenedTypeRelation {

    FieldLinkDefinition[][] fieldLinks;

    public FlattenedTypeRelation(FlattenedType flattenedType, List<FieldLinkDefinition> fls) {
        Map<Integer, List<FieldLinkDefinition>> groupedByOriginFD =
                fls.stream()
                .collect(Collectors.groupingBy(fl ->
                        fl.getOriginFD().getId())
                );

        fieldLinks = new FieldLinkDefinition[flattenedType.getFieldsReverse().length][];
        for(short i = 0; i < fieldLinks.length; i++) {
            for(FieldDefinition fd : flattenedType.getFieldsReverse()[i]) {
                List<FieldLinkDefinition> list = groupedByOriginFD.get(fd.getId());
                if (list != null) {
                    fieldLinks[i] = list.toArray(new FieldLinkDefinition[0]);
                }
            }
        }
    }

    public void followLinks(Direction direction, Obj relatedObj, Field field) {
        FieldLinkDefinition[] fls = fieldLinks[field.getId()];
        if(fls != null) {
            for (FieldLinkDefinition fl : fls) {
                direction.transmit(
                        field,
                        fl,
                        relatedObj
                );
            }
        }
    }
}
